### PR TITLE
add host exerciser lpbk sample

### PR DESCRIPTION
**Pull request by anandaravuri**
-host exercise lpbk  sample   tests loop back afu in different modes

Signed-off-by: anandaravuri <ananda.ravuri@intel.com>

### CODE REVIEW

File: samples/host_exerciser/host_exerciser.h [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+using test_command = opae::afu_test::command;
+
+class host_exerciser : public test_afu {
+public:
+	host_exerciser(const char *afu_id = AFU_ID)
+  : test_afu("host_exerciser", afu_id)
+  , count_(1)
+  {
+    app_.add_option("-m,--mode", mode_str_, "mode {lpbk,read, write, trput}")->default_val("lpbk");
+    app_.add_option("-r,--read", read_str_, "read request type{rdline_s, rdline_l,mixed}")->default_val("rdline_s");
```
Comment (r-rojo) [inline samples/host_exerciser/host_exerciser.h:338]: similar comment as for line 338

File: samples/host_exerciser/host_exerciser.h [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+    app_.add_option("-w,--write", write_str_, "write request type {wrline_m, wrline_l}")->default_val("wrline_m");
```
Comment (r-rojo) [inline samples/host_exerciser/host_exerciser.h:339]: similar comment as for line 338

File: samples/host_exerciser/host_exerciser.h [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+    app_.add_option("-d,--delay", delay_, "Enables random delay insertion between requests")->default_val(false);
+    app_.add_option("--multi-cl", multi_cl_, "multi Cache lineone of {0, 1, 3}")->default_val(0);
```
Comment (r-rojo) [inline samples/host_exerciser/host_exerciser.h:341]: similar comment as for line 338 except we wouldn't use a transformer.
Here instead we could just use CLI::IsMember.
```c++
  app_.add_option("--multi-cl", multi_cl_, "multi Cache lineone of {0, 1, 3}")->default_val(0)
    ->check(CLI::IsMember({0, 1, 3}));
```

Also, are the multi cli values are 0, 1, and 3? I thought it was 0, 1, and 2.

File: samples/host_exerciser/host_exerciser_lpbk.h [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+  void host_exerciser_errors()
+  {
+    csr_error afu_error(0);
+    csr_error1 afu_error1(0);
+
+    afu_error.value = he_afu_->read64(CSR_STATUS0);
+    afu_error1.value = he_afu_->read64(CSR_STATUS1);
+
+     std::cout << "Host Exerciser Error:" << afu_error.error << std::endl;
```
Comment (r-rojo) [inline samples/host_exerciser/host_exerciser_lpbk.h:73]: seems like spacing is off

File: samples/host_exerciser/host_exerciser_lpbk.h [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    return 0;
+  }
+
+  virtual int run(test_afu *afu, CLI::App *app)
+  {
+    (void)app;
+
+    auto d_afu = dynamic_cast<host_exerciser*>(afu);
+   he_afu_ = dynamic_cast<host_exerciser*>(afu);
+   uint32_t input_cmds;
```
Comment (r-rojo) [inline samples/host_exerciser/host_exerciser_lpbk.h:169]: spacing seems like it's off

File: samples/host_exerciser/host_exerciser_lpbk.h [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+
+	auto ret = parse_input_options(afu,&input_cmds);
```
Comment (r-rojo) [inline samples/host_exerciser/host_exerciser_lpbk.h:171]: We can possibly use a member variable of type csr_cfg, and pass appropriate member of this in validators for CLI11.
If we do this, we can probably skip this step.

File: samples/host_exerciser/host_exerciser.h [hunk 7]
```diff
@@ -0,0 +1,489 @@
...
+
+// TBD
+union dfh_cap_guid {
+	enum {
+		offset = DFH_CAPABILITY_GUID
+	};
+
+	dfh_cap_guid(uint8_t* v)
+	{
+		for (int i = 0; i < 24; i++) {
```
Comment (tswhison) [inline samples/host_exerciser/host_exerciser.h:136]: sizeof(value) to be safe?

File: samples/host_exerciser/host_exerciser.h [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+
+
+union dfh_cap_init {
+	enum {
+		offset = DFH_CAPABILITY_INT
+	};
+
+	dfh_cap_init(uint8_t* v)
+	{
+		for (int i = 0; i < 20; i++) {
```
Comment (tswhison) [inline samples/host_exerciser/host_exerciser.h:158]: sizeof(value) to be safe?

File: samples/host_exerciser/host_exerciser.h [hunk 9]
```diff
@@ -0,0 +1,489 @@
...
+			memcpy(&value[i], v, 1);
+			v++;
+		}
+	}
+	uint8_t  value[20];
+	struct {
+		uint32_t IntVectorLength ;
```
Comment (tswhison) [inline samples/host_exerciser/host_exerciser.h:165]: s/IntVectorLength ;/IntVectorLength;/

File: samples/host_exerciser/host_exerciser.h [hunk 10]
```diff
@@ -0,0 +1,489 @@
...
+		uint32_t IntVectorStart;
+		uint32_t IntControl ;
```
Comment (tswhison) [inline samples/host_exerciser/host_exerciser.h:167]: s/IntControl ;/IntControl;/

File: samples/host_exerciser/host_exerciser.h [hunk 11]
```diff
@@ -0,0 +1,489 @@
...
+		uint32_t IntFlags ;
```
Comment (tswhison) [inline samples/host_exerciser/host_exerciser.h:168]: s/IntFlags ;/IntFlags;/

File: samples/host_exerciser/host_exerciser.h [hunk 12]
```diff
@@ -0,0 +1,489 @@
...
+    app_.add_option("-m,--mode", mode_str_, "mode {lpbk,read, write, trput}")->default_val("lpbk");
+    app_.add_option("-r,--read", read_str_, "read request type{rdline_s, rdline_l,mixed}")->default_val("rdline_s");
+    app_.add_option("-w,--write", write_str_, "write request type {wrline_m, wrline_l}")->default_val("wrline_m");
+    app_.add_option("-d,--delay", delay_, "Enables random delay insertion between requests")->default_val(false);
+    app_.add_option("--multi-cl", multi_cl_, "multi Cache lineone of {0, 1, 3}")->default_val(0);
+    app_.add_option("--ccont", c_cont_, "Configures test rollover or test termination")->default_val(false);
+  }
+
+
+  virtual int run(CLI::App *app, test_command::ptr_t test) override
```
Comment (tswhison) [inline samples/host_exerciser/host_exerciser.h:346]: Can this be inherited from test_afu?

File: samples/host_exerciser/host_exerciser_lpbk.h [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+  void host_exerciser_status()
+  {
+    csr_status0 afu_status0(0);
+    csr_status1 afu_status1(0);
+
+    afu_status0.value = he_afu_->read64(CSR_STATUS0);
+    afu_status1.value = he_afu_->read64(CSR_STATUS1);
+
+    std::cout << "Host Exerciser numReads:"<< afu_status0.numReads << std::endl;
+     std::cout << "Host Exerciser numReads:" << afu_status0.numWrites << std::endl;
```
Comment (tswhison) [inline samples/host_exerciser/host_exerciser_lpbk.h:59]: spacing is off.

File: samples/host_exerciser/host_exerciser.h (hunk 7 above)
Comment (r-rojo) [reply samples/host_exerciser/host_exerciser.h:136]: Or maybe use std::copy or memcpy into the entire value?
```c++
dfh_cap_guid(uint8_t *v)
{
  std::copy(v, v+sizeof(value), value);
}
```

File: samples/host_exerciser/host_exerciser_lpbk.h (hunk 13 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser_lpbk.h]: fixed

File: samples/host_exerciser/host_exerciser.h (hunk 11 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: Fixed

File: samples/host_exerciser/host_exerciser.h (hunk 10 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: Fixed

File: samples/host_exerciser/host_exerciser.h (hunk 9 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: Fixed

File: samples/host_exerciser/host_exerciser.h (hunk 8 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: defined structure  dfh_cap_init_s and remove memory copy. fixed

File: samples/host_exerciser/host_exerciser.h (hunk 7 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: fixed

File: samples/host_exerciser/host_exerciser_lpbk.h (hunk 6 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser_lpbk.h]: Added csr_cfg as member variable

File: samples/host_exerciser/host_exerciser_lpbk.h (hunk 5 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser_lpbk.h]: fixed

File: samples/host_exerciser/host_exerciser_lpbk.h (hunk 4 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser_lpbk.h:79]: fixed

File: samples/host_exerciser/host_exerciser.h (hunk 3 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: fixed

File: samples/host_exerciser/host_exerciser.h (hunk 2 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: fixed

File: samples/host_exerciser/host_exerciser.h (hunk 1 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: fixed

File: samples/host_exerciser/host_exerciser.h [hunk 14]
```diff
@@ -0,0 +1,489 @@
...
+using test_afu = opae::afu_test::afu;
+using test_command = opae::afu_test::command;
+
+class host_exerciser : public test_afu {
+public:
+	host_exerciser(const char *afu_id = AFU_ID)
+  : test_afu("host_exerciser", afu_id)
+  , count_(1)
+  {
+    app_.add_option("-m,--mode", mode_str_, "mode {lpbk,read, write, trput}")->default_val("lpbk");
```
[the remark this answers is not in the document]
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: fixed

File: samples/host_exerciser/host_exerciser.h [hunk 15]
```diff
@@ -334,15 +327,29 @@ class host_exerciser : public test_afu {
...
+    // Delay
     app_.add_option("-d,--delay", delay_, "Enables random delay insertion between requests")->default_val(false);
-    app_.add_option("--multi-cl", multi_cl_, "multi Cache lineone of {0, 1, 3}")->default_val(0);
+
+    // Multi Cache line
+    app_.add_option("--multi-cl", multi_cl_, "multi Cache lineone of {0, 1, 3}")
+         ->transform(CLI::IsMember(std::set<int>({ 0,1,3 })))->default_val(0);
+
+    // Configures test rollover
     app_.add_option("--ccont", c_cont_, "Configures test rollover or test termination")->default_val(false);
```
[the remark this answers is not in the document]
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: fixed

File: samples/host_exerciser/host_exerciser.h [hunk 16]
```diff
@@ -334,15 +327,29 @@ class host_exerciser : public test_afu {
...
+    // Write
+    app_.add_option("-w,--write", write_, "write request type {wrline_m, wrline_l}")
+        ->transform(CLI::CheckedTransformer(writes))->default_val("wrline_m");
+
+    // Delay
     app_.add_option("-d,--delay", delay_, "Enables random delay insertion between requests")->default_val(false);
-    app_.add_option("--multi-cl", multi_cl_, "multi Cache lineone of {0, 1, 3}")->default_val(0);
+
+    // Multi Cache line
+    app_.add_option("--multi-cl", multi_cl_, "multi Cache lineone of {0, 1, 3}")
```
[the remark this answers is not in the document]
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: fixed

File: samples/host_exerciser/host_exerciser.h [hunk 17]
```diff
@@ -334,15 +327,29 @@ class host_exerciser : public test_afu {
...
+    // Read
+    app_.add_option("-r,--read", read_, "read request type{rdline_s, rdline_l,mixed}")
+       ->transform(CLI::CheckedTransformer(reads))->default_val("rdline_s");
+
+    // Write
+    app_.add_option("-w,--write", write_, "write request type {wrline_m, wrline_l}")
+        ->transform(CLI::CheckedTransformer(writes))->default_val("wrline_m");
+
+    // Delay
     app_.add_option("-d,--delay", delay_, "Enables random delay insertion between requests")->default_val(false);
```
[the remark this answers is not in the document]
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: fixed

File: samples/host_exerciser/host_exerciser.h [hunk 18]
```diff
@@ -334,15 +327,29 @@ class host_exerciser : public test_afu {
...
+    // Mode
+    app_.add_option("-m,--mode", mode_, "mode {lpbk,read, write, trput}")
+      ->transform(CLI::CheckedTransformer(modes))->default_val("lpbk");
+
+    // Read
+    app_.add_option("-r,--read", read_, "read request type{rdline_s, rdline_l,mixed}")
+       ->transform(CLI::CheckedTransformer(reads))->default_val("rdline_s");
+
+    // Write
+    app_.add_option("-w,--write", write_, "write request type {wrline_m, wrline_l}")
```
[the remark this answers is not in the document]
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: fixed

File: samples/host_exerciser/host_exerciser.h [hunk 19]
```diff
@@ -334,15 +327,29 @@ class host_exerciser : public test_afu {
...
   {
-    app_.add_option("-m,--mode", mode_str_, "mode {lpbk,read, write, trput}")->default_val("lpbk");
-    app_.add_option("-r,--read", read_str_, "read request type{rdline_s, rdline_l,mixed}")->default_val("rdline_s");
-    app_.add_option("-w,--write", write_str_, "write request type {wrline_m, wrline_l}")->default_val("wrline_m");
+    // Mode
+    app_.add_option("-m,--mode", mode_, "mode {lpbk,read, write, trput}")
+      ->transform(CLI::CheckedTransformer(modes))->default_val("lpbk");
+
+    // Read
+    app_.add_option("-r,--read", read_, "read request type{rdline_s, rdline_l,mixed}")
```
[the remark this answers is not in the document]
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: fixed

File: samples/host_exerciser/host_exerciser.h [hunk 20]
```diff
@@ -334,15 +327,29 @@ class host_exerciser : public test_afu {
   : test_afu("host_exerciser", afu_id)
   , count_(1)
   {
-    app_.add_option("-m,--mode", mode_str_, "mode {lpbk,read, write, trput}")->default_val("lpbk");
-    app_.add_option("-r,--read", read_str_, "read request type{rdline_s, rdline_l,mixed}")->default_val("rdline_s");
-    app_.add_option("-w,--write", write_str_, "write request type {wrline_m, wrline_l}")->default_val("wrline_m");
+    // Mode
+    app_.add_option("-m,--mode", mode_, "mode {lpbk,read, write, trput}")
```
[the remark this answers is not in the document]
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: fixed

File: doc/src/fpga_tools/host_exerciser/host_exerciser.md [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+host exerciser tool time out, by default time out 60000
+
+`-m,--mode`
+
+host exerciser modes are supported lpbk,read, write, trput
+
+
+`--cl`
+
+Number of CLs per request 1, 2, 3, 4.
```
Comment (r-rojo) [inline doc/src/fpga_tools/host_exerciser/host_exerciser.md:90]: s/CL/cachelines/

File: samples/host_exerciser/host_exerciser.py [hunk 22]
```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python3
```
Comment (r-rojo) [inline samples/host_exerciser/host_exerciser.py:1]: Why do we need this file?

**Review by r-rojo — COMMENTED** [review]
Don't forget to change copyright year to 2021 for new files and 2020-2021 for files included last year but changed this year.

File: doc/src/fpga_tools/host_exerciser/host_exerciser.md [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+
+open FPGA PCIe resource in shared mode
+
+`-t,--timeout`
+
+host exerciser tool time out, by default time out 60000
+
+`-m,--mode`
+
+host exerciser modes are supported lpbk,read, write, trput
```
Comment (tswhison) [inline doc/src/fpga_tools/host_exerciser/host_exerciser.md:85]: s/lpbk,read/lpbk, read/

File: doc/src/fpga_tools/host_exerciser/host_exerciser.md [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+
+
+`--cl`
+
+Number of CLs per request 1, 2, 3, 4.
+
+
+`--ccont`
```
Comment (tswhison) [inline doc/src/fpga_tools/host_exerciser/host_exerciser.md:93]: Should this be --cont?

File: samples/host_exerciser/CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+        ${libuuid_LIBRARIES}
+    COMPONENT samplebin
+)
+
+
+set(host_exerciser_src host_exerciser.py)
+
+install(PROGRAMS ${host_exerciser_src}
+    DESTINATION bin
+    COMPONENT samplebin)
```
Comment (tswhison) [inline samples/host_exerciser/CMakeLists.txt:43]: Needs a newline at end of file.

File: samples/host_exerciser/host_exerciser.h [hunk 26]
```diff
@@ -0,0 +1,449 @@
...
+    // Mode
+    app_.add_option("-m,--mode", he_modes_, "host exerciser mode {lpbk,read, write, trput}")
+      ->transform(CLI::CheckedTransformer(he_modes))->default_val("lpbk");
+
+    // Cache line
+    app_.add_option("--cl", he_req_cl_len_, "number of CLs per request{cl_1, cl_2, cl_3, cl_4}")
+       ->transform(CLI::CheckedTransformer(he_req_cl_len))->default_val("cl_1");
+
+    // Configures test rollover or test termination
+    app_.add_option("--ccont", he_ccont_, "Configures test rollover or test termination")->default_val(false);
```
Comment (tswhison) [inline samples/host_exerciser/host_exerciser.h:306]: Oh, it is --ccont.  Just checking, is that what you intended?

File: doc/src/fpga_tools/host_exerciser/host_exerciser.md (hunk 21 above)
Comment (anandaravuri) [reply doc/src/fpga_tools/host_exerciser/host_exerciser.md]: fixed

File: samples/host_exerciser/host_exerciser.py (hunk 22 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.py]: fixed

File: doc/src/fpga_tools/host_exerciser/host_exerciser.md (hunk 23 above)
Comment (anandaravuri) [reply doc/src/fpga_tools/host_exerciser/host_exerciser.md]: fixed

File: doc/src/fpga_tools/host_exerciser/host_exerciser.md (hunk 24 above)
Comment (anandaravuri) [reply doc/src/fpga_tools/host_exerciser/host_exerciser.md]: fixed

File: samples/host_exerciser/CMakeLists.txt (hunk 25 above)
Comment (anandaravuri) [reply samples/host_exerciser/CMakeLists.txt]: fixed

File: samples/host_exerciser/host_exerciser.h (hunk 26 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.h]: fixed

File: samples/host_exerciser/host_exerciser.py (hunk 22 above)
Comment (anandaravuri) [reply samples/host_exerciser/host_exerciser.py]: Multi PF & VF test tool, used to test HE-LPBK and HE-ME and HSSI

**Pull request closed without merging**